### ISSUE TITLE
tuple field access via a json path

### ISSUE THREAD

**Issue by kostja** [tarantool/tarantool#1285]
Extend the tuple API in Lua and in C to make it possible to access tuple fields via a json path, for example:

`tuple.path.to.field`

or 

`tuple["path.to.field"]`
## Implementation

The implementation has to handle two cases independently: when a tuple is flattened (see gh-1256), the field should be accessed by its numeric index, taken from the mapping. If the tuple is not flattened, the field data (in MsgPack) is parsed/searched until a matching path is found.

It is necessary to support the second case to provide seamless user experience for tuple-space (i.e. spaces containing tuples with arrays of fields, for such tuples the flattening is off) and document-space (spaces containing flattened documents).

**Comment by Gerold103** [comment]
The patch is already done for non-unicode characters. The proposal is to use wchar instead of char in the json path parser to correctly process paths with hieroglyphs and other strange symbols.

**Comment by Gerold103** [comment]
"What about flattening/unflattening support?" you can ask, but I can answer, that the ticket about documents support is rather blocked by this ticket than this ticker is blocked by documents. We can not implement documents support while can not 1) access tuple fields by JSON, 2) update tuple fields by JSON, 3) index tuples by JSON.

**Comment by Gerold103** [comment]
@TarantoolBot docu

**Comment by TarantoolBot** [comment]
@Gerold103: Invalid request type.

**Comment by Gerold103** [comment]
@TarantoolBot document
no title

**Comment by TarantoolBot** [comment]
@Gerold103: Title not found.

**Comment by Gerold103** [comment]
@TarantoolBot doc

**Comment by TarantoolBot** [comment]
@Gerold103: Invalid request type.

**Comment by Gerold103** [comment]
@TarantoolBot document
Title: Document tuple data access by JSON path
Until this moment in Lua a tuple data could be accessed in 3 ways:
- get field by field number, decode into Lua,
- get field by field name, decode into Lua
- decode entire tuple into Lua.

It was impossible to decode into Lua only a part of the field to avoid Lua garbage creating. For example, consider the tuple: `{1, 2, {key1 = value1, key2 = {key21 = {key211 = {key2111, key2112}}}}}` with field names `field1`, `field2`, `field3`. To get `key2112` it is necessary to decode into Lua the entire 3-th field, including `key1`, `value1`, `key2`, `key21`, `key211`, `key2111` using the syntax `key2112 = tuple.field3.key2.key21.key211[2]`. Now the one can use the following syntax: `key2112 = tuple["field3.key2.key21.key211[2]"]`. As you can see the difference is in placing all the path into quotes and brackets `["..."]`. The Tarantool goes through this path and MessagePack tuple body, gets the needed tuple part and decodes into Lua only it.

The path must be valid JSON path with one exception - in Tarantool a path can start with `.`. For example: `tuple[".field3..."]` - it is done to lay emphasis that the JSON path here is just a suffix for the tuple.

At the same time tuple field names still work: `tuple["field1"]`, `tuple.field2`.

If the field name looks like JSON path, for example: `my.field.name`, then it works too. The path at first is checked to be a real JSON path, and if nothing is found, then the entire path is considered as a field name. To combine such names and path elements the one can use `["..."]`. For example: `tuple["['my.field.name'].key1.key2.['another.key.in.msgpack_map']"]`.

**Comment by TarantoolBot** [comment]
@Gerold103: Accept.
